### PR TITLE
Google Slides Workload

### PR DESCRIPTION
**Pull request by muendelezaji**

### CODE REVIEW

File: wlauto/workloads/googleslides/__init__.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+
+def log_method(workload, name):
+    workload.logger.info('===== {}() ======'.format(name))
+
+class GoogleSlides(AndroidUiAutoBenchmark):
+
+    name = 'googleslides'
+    package = 'com.google.android.apps.docs.editors.slides'
+    description = 'Creates a Google Slides presentation with some commonly used features'
```
Comment (jimboatarm) [inline wlauto/workloads/googleslides/__init__.py]: Can you fill out the description with all of the steps performed by the workload?  We plan to take this text directly into the final report so it should follow the format established in previous workloads.

File: wlauto/workloads/googleslides/__init__.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+    instrumentation_log = '{}_instrumentation.log'.format(name)
+
+    def __init__(self, device, **kwargs):
+        super(GoogleSlides, self).__init__(device, **kwargs)
+        self.run_timeout = 300
+        self.output_file = path.join(self.device.working_directory, self.instrumentation_log)
+        self.local_dir = self.dependencies_directory
+        # Android downloads folder
+        self.device_dir = path.join(self.device.working_directory, '..', 'Download')
```
Comment (jimboatarm) [inline wlauto/workloads/googleslides/__init__.py:95]: Why the downloads folder?   Does this guarantee the document turns up in a known place?  My experience is that you can't trust pushing files to Downloads and know they will appear within an application.

File: wlauto/workloads/googleslides/__init__.py (hunk 1 above)
Comment (muendelezaji) [inline wlauto/workloads/googleslides/__init__.py]: OK, will do.

File: wlauto/workloads/googleslides/__init__.py (hunk 2 above)
Comment (muendelezaji) [inline wlauto/workloads/googleslides/__init__.py:95]: In the 4 devices I was able to get my hands on (if only momentarily) to test on, this seemed to guarantee that the file will reliably found — `Download` is the default folder for the `Open` action in Slides. The search function only includes online files, or offline files previously opened by the app, so can't be used to locate the file first time.

Otherwise, pushing it to WA's working directory meant trying to navigate up/back in the hierarchy to try and find it. This was not as reliable as the appearance of the file picker differed by device.

I'd be happy to tweak the implementation if we find that the `Download` folder is in fact less reliable.

Comment (jimboatarm) [inline wlauto/workloads/googleslides/__init__.py:95]: Ok.  Thanks for testing this so comprehensively.  We'll verify it here once it is merged.

File: wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java [hunk 3]
```diff
@@ -0,0 +1,469 @@
...
+
+    public UiObject clickView(FindByCriteria criteria, String matching, boolean wait) throws Exception {
+        return clickView(criteria, matching, null, wait);
+    }
+
+    public UiObject clickView(FindByCriteria criteria, String matching, String clazz) throws Exception {
+        return clickView(criteria, matching, clazz, false);
+    }
+
+    public UiObject clickView(FindByCriteria criteria, String matching, String clazz, boolean wait) throws Exception {
```
Comment (jimboatarm) [inline wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java]: Is 'View' the best name to use?  These methods are working with UiObjects not Views.  We should consider changing these names

File: wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java [hunk 4]
```diff
@@ -0,0 +1,469 @@
...
+        while (++slideIndex < slideCount) {
+            testTag = "slideshow_next_" + slideIndex;
+            startDumpsys(ACTIVITY_SLIDES);
+            slideTimer = new Timer();
+            slideTimer.start();
+            uiDeviceSwipeHorizontal(centerX + centerX/2, centerX - centerX/2, centerY);
+            slideTimer.end();
+            results.put(testTag, slideTimer);
+            endDumpsys(ACTIVITY_SLIDES, testTag);
+            sleep(1);
```
Comment (jimboatarm) [inline wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java]: Why are all of these sleeps necessary?

File: wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java [hunk 5]
```diff
@@ -0,0 +1,469 @@
...
+        UiObject decreaseFont = getViewByDesc("Decrease text");
+        repeatClickView(decreaseFont, 20);
+        getUiDevice().pressBack();
+
+        // get image from gallery and insert
+        insertSlide("Title Only");
+        clickView(BY_DESC, "Insert");
+        clickView(BY_TEXT, "Image", true);
+        clickView(BY_TEXT, "Recent");
+        try {
```
Comment (jimboatarm) [inline wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java:287]: I haven't got a device to hand - do we need to push test images to be sure that the same image is used each type of handset?

File: wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java [hunk 6]
```diff
@@ -0,0 +1,469 @@
...
+    public void enterTextInSlide(String viewName, String textToEnter) throws Exception {
+        UiObject view = getViewByDesc(viewName);
+        view.click();
+        view.setText(textToEnter);
+        try {
+            clickView(BY_DESC, "Done");
+        } catch (UiObjectNotFoundException e) {
+            clickView(BY_ID, "android:id/action_mode_close_button");
+        }
+        SystemClock.sleep(200);
```
Comment (jimboatarm) [inline wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java]: Why is this necessary?

File: wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java [hunk 7]
```diff
@@ -0,0 +1,469 @@
...
+
+    public void saveDocument(String docName) throws Exception {
+        timer = new Timer();
+        timer.start();
+        clickView(BY_TEXT, "SAVE");
+        clickView(BY_TEXT, "Device");
+        timer.end();
+        results.put("save_dialog1", timer);
+        // Allow SD card access if requested
+        UiObject permissionView = new UiObject(new UiSelector().textContains("Allow Slides"));
```
Comment (jimboatarm) [inline wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java]: Is this necessary now that you have found a way to set permissions with your fix from a previous commit?

File: wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java (hunk 4 above)
Comment (muendelezaji) [inline wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java]: This is so that we get a perceptible wait-time before the next slide (i.e. real user behaviour). Otherwise it just swipes through the slides too quickly.
Edit:
Note that they are not measured in the per-slide timings, and introduce a systematic 1 second per slide transition in the overall timings.

File: wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java [hunk 8]
```diff
@@ -0,0 +1,469 @@
...
+        timer = new Timer();
+        timer.start();
+        UiObject filename = getViewById(PACKAGE_ID + "file_name_edit_text");
+        filename.clearTextField();
+        filename.setText(docName);
+        clickView(BY_TEXT, "Save", CLASS_BUTTON);
+        timer.end();
+        results.put("save_dialog2", timer);
+        // Overwrite if prompted
+        UiObject overwriteView = new UiObject(new UiSelector().textContains("already exists"));
```
Comment (jimboatarm) [inline wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java:359]: How would we end up with a document of the same name?  If the test always starts from the same initial conditions this shouldn't be possible.

File: wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java [hunk 9]
```diff
@@ -0,0 +1,469 @@
...
+        SystemClock.sleep(200);
+    }
+
+    public void saveDocument(String docName) throws Exception {
+        timer = new Timer();
+        timer.start();
+        clickView(BY_TEXT, "SAVE");
+        clickView(BY_TEXT, "Device");
+        timer.end();
+        results.put("save_dialog1", timer);
```
Comment (jimboatarm) [inline wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java]: change to save_dialog_1.  It makes it easier to read and potentially parse the results in iPython later.  We should adopt this style for all cases where we have multiple similar actions that need numerating.  Do the same for delete below.

File: wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java [hunk 10]
```diff
@@ -0,0 +1,469 @@
...
+        timer.end();
+        results.put("delete_dialog2", timer);
+        sleep(1);
+    }
+
+    public void repeatClickView(UiObject view, int repeat) throws Exception {
+        if (repeat < 1 || !view.isClickable()) return;
+        while (repeat-- > 0) {
+            view.click();
+            SystemClock.sleep(50); // in order to register as separate click
```
Comment (jimboatarm) [inline wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java]: Does 50 work on all devices?  It seems an arbitrary magic number.

File: wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java [hunk 11]
```diff
@@ -0,0 +1,469 @@
...
+            deleteButton = getUiObjectByText("Remove", CLASS_BUTTON);
+        } catch (UiObjectNotFoundException e) {
+            deleteButton = getUiObjectByText("Ok", CLASS_BUTTON);
+        }
+        deleteButton.clickAndWaitForNewWindow();
+        timer.end();
+        results.put("delete_dialog2", timer);
+        sleep(1);
+    }
+
```
Comment (jimboatarm) [inline wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java:388]: I'm taking it that everything from here down could go into the base classes.  I'm happy for you to do that as part of this change, please.

File: wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java (hunk 5 above)
Comment (muendelezaji) [inline wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java:287]: As with the default folder, to keep the test simple we just select the most recent image from the gallery (regardless of what folder it is located in). It performed more reliably than pushing an image then trying to find it in the picker.
The image did not seem to affect the results, and in the rare case that there will be none at all, the test does not fall over.

File: wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java (hunk 4 above)
Comment (jimboatarm) [inline wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java]: If there is a legitimate reason then explain each case with a comment.  We're all sensitive to magic numbers and sleeps when we see them in code.  I'll have a look on a real device and see what this looks like.  As we are not measuring anything during these sleeps it isn't adding much to the test apart from visual appeal.  We won't get any results we can compare.  It's just adding time to each test iteration.

File: wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java (hunk 6 above)
Comment (muendelezaji) [inline wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java]: Introduces a small wait time for stability (not measured). On some devices, keyboard pops up when entering text, and takes a finite amount of time to disappear after clicking Done. In these cases, trying to find a view immediately after entering text leads to `UiObjectNotFoundException`

File: wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java (hunk 9 above)
Comment (muendelezaji) [inline wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java]: Will do

File: wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java (hunk 8 above)
Comment (muendelezaji) [inline wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java:359]: Should not happen under normal circumstances. But when the test sometimes failed for whatever reason, the created file would not be deleted and thus be on the device during the next run.

Note that part of the test is to delete the file in-app, so it isn't removed as part workload teardown. One scenario where I've seen this happen is when long-click on the document list fails.

It is not included in timings, so shouldn't skew results.

File: wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java (hunk 7 above)
Comment (muendelezaji) [inline wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java]: I suppose not anymore. Had to have it as the functionality was not in master at the time.

File: wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java (hunk 4 above)
Comment (muendelezaji) [inline wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java]: Will add a comment

File: wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java (hunk 10 above)
Comment (muendelezaji) [inline wlauto/workloads/googleslides/uiauto/src/com/arm/wlauto/uiauto/UiAutomation.java]: Seemed like a sensible default from some trial & error and a few sources on the internet.
Specifically, 50ms per iteration shouldn't increase the test time significantly e.g. current usage is 1 second (20 clicks \* 50ms). I suspect it can even be lowered, though I could not confirm this.
It can be raised/lowered it if found to be too short/long.

**Comment by muendelezaji** [comment]
Changes after review:
- [x] no longer require permission dialog handling
- [x] replace magic numbers with named constants
- [x] better names for save/delete dialog test tags
- [x] comments to explain some design choices
- [x] rename `getView`/`clickView` for consistency with existing methods
- [x] move common/convenience functions to base classes for reuse
